calculator_test.go

```diff
@@ -2,23 +2,192 @@ package calculator_test
+		{"one fractional and one whole number", 5.4, 2, 7.4},
+		{"two fractional numbers", 2.3, 4.3, 6.6},
+		{"two fractional numbers equaling a whole number", 2.3, 3.7, 6},
+		{"two negative fractional numbers", -1.5, -2.5, -4},
+	}
+
+	t.Parallel()
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			want := c.want
+			got := calculator.Add(c.a, c.b)
+			if want != got {
+				t.Errorf(c.name+": want %f, got %f", want, got)
+			}
+		})
+	}
+}
+
+func TestSubtract(t *testing.T) {
+	cases := []testCase{
+		{"two positive numbers", 5, 1, 4},
+		{"two negative numbers", -5, -1, -4},
+		{"two negative numbers equaling zero", -1, -1, 0},
+		{"a fractional number from a whole number", 5.0, 1.5, 3.5},
+		{"two fractional numbers", 2.3, 4.3, -2},
+		{"two fractional numbers equaling a whole number", 2.5, 1.5, 1},
+		{"two negative fractional numbers", -2.5, -1.5, -1},
+		{"a negative number from a positive number", 5, -3, 8},
+		{"a postitive number from a negative number", -4, 2, -6},
+	}
+
 	t.Parallel()
-	var want float64 = 4
-	got := calculator.Add(2, 2)
-	if want != got {
-		t.Errorf("want %f, got %f", want, got)
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			want := c.want
+			got := calculator.Subtract(c.a, c.b)
+			if want != got {
+				t.Errorf(c.name+": want %f, got %f", want, got)
+			}
+		})
 	}
 }
 
-// func TestSubtract(t *testing.T) {
-// 	t.Parallel()
-// 	var want float64 = 2
-// 	got := calculator.Subtract(4, 2)
-// 	if want != got {
-// 		t.Errorf("want %f, got %f", want, got)
-// 	}
-// }
+func TestMultiply(t *testing.T) {
+	cases := []testCase{
+		{"two positive numbers", 2, 2, 4},
+		{"two negative numbers", -2, -2, 4},
+		{"a postive number by zero", 1, 0, 0},
+		{"a fractional by a whole number", 5.4, 2, 10.8},
+		{"two fractional numbers", 2.6, 5.3, 13.78},
+		{"two fractional numbers equaling a whole number", 1.5, 4, 6},
+		{"two negative fractional numbers", -1.5, -2.5, 3.75},
+		{"a postive number by a negative number", -5, 3, -15},
+	}
+
+	t.Parallel()
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			want := c.want
+			got := calculator.Multiply(c.a, c.b)
+			if want != got {
+				t.Errorf(c.name+": want %f, got %f", want, got)
+			}
+		})
+	}
+}
+
+func TestDivide(t *testing.T) {
+	cases := []testCaseWithErr{
+		{"two positive numbers", 4, 2, false, 2},
+		{"two negative numbers", -4, -2, false, 2},
+		{"a positive number by a negative number", 4, -2, false, -2},
+		{"a negative number by a postive number", -4, 2, false, -2},
+		{"two fractional positive numbers", 4.2, 2.1, false, 2},
+		{"two fractional negative numbers equaling a whole number", -4.2, -2.1, false, 2},
+		{"a positive number by zero", 4, 0, true, 0},
+	}
+
+	t.Parallel()
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			want := c.want
+			got, err := calculator.Divide(c.a, c.b)
+			if err != nil && !c.errExpected {
+				t.Errorf(c.name+": wanted %f got an error: %v ", want, err)
```

Comment:
```suggestion
				t.Fatalf(c.name+": wanted %f got an error: %v ", want, err)
```

We should bail, not fail, here, because if the error status is not right, we already know that the data value is invalid.

Reply:
Good point.  Reading what Fatalf does, it seems like this would terminate the whole test when I got to that case.  The other tests will still run, but not the other cases in this test.  Are there times when it would be preferable to still Errorf so the rest of the cases run?

Reply:
Fatal bails out of the current test _or subtest_. So in a test case being executed by t.Run(), calling Fatal exits this test case and goes on to the next case.

Essentially, Fatal says "for this test case, it's not worth doing any more checks". There's a slightly more detailed discussion in the book, in the section "Failing and Bailing": https://bitfieldconsulting.com/books/fundamentals